returnn/tf/layers/rec.py

```diff
@@ -1082,6 +1085,11 @@ def collect_exception(cls, layer_name):
           else:
             out = StringIO()
             better_exchook.better_exchook(exc_type, value, tb, file=out)
+            if isinstance(value, fail_directly_exception_types):
+              raise Exception(
```

Comment:
Why actually do you throw a new exception?
Why not simply reraise the original exception, i.e. just `raise` or `raise value`?

Reply:
Is it possible to add text after the Exception message then?

Reply:
Why do you want to edit it? It already has all the information. And when you pass it through, it will end up to the user as the final failing exception.

Reply:
The way it is now it will end up as the final exception, it is really just about the sentence...

```
[...]
    line: raise DataNotFound(
            "ExternData: unknown key %r. available keys: %s. config: %s" % (
              name, list(self.data.keys()), config_extern_data))
    locals:
      DataNotFound = <global> <class 'returnn.tf.network.DataNotFound'>
      name = <local> 'decoder_cells', len = 13
      list = <builtin> <class 'list'>
      self = <local> <ExternData data={'classes': Data(name='classes', dtype='int32', sparse=True, dim=54, available_for_inference=False, batch_shape_meta=[B]), 'data': Data(name='data', batch_shape_meta=[B,F|80]), 'speaker_label': Data(name='speaker_label', dtype='int32', sparse=True, dim=251, available_for_inferenc...
      self.data = <local> {'classes': Data(name='classes', dtype='int32', sparse=True, dim=54, available_for_inference=False, batch_shape_meta=[B]), 'data': Data(name='data', batch_shape_meta=[B,F|80]), 'speaker_label': Data(name='speaker_label', dtype='int32', sparse=True, dim=251, available_for_inference=False, batch_sh..., len = 7
      self.data.keys = <local> <built-in method keys of dict object at 0x7f082e1105c0>
      config_extern_data = <local> '<unknown>', len = 9
DataNotFound: ExternData: unknown key 'decoder_cells'. available keys: ['classes', 'data', 'speaker_label', 'ctx', 'padded_data_target', 'windowed_data_target', 'stop_token_target']. config: <unknown>

Template construction failed with non-allowed Exception while constructing layer: data:decoder_cells

Reply:
All unhandled exceptions are always "non-allowed".
I still don't see how just passing through the exception (i.e. not handling it) is not just the same (but more clean).

Reply:
Okay, this is not super important for me so I can remove that.

Still, I need to figure out how to capture Exceptions and make them more verbose without raising a different exception / using `from` (because depending on the stack this makes it not nice to read). But we can discuss this offline.